README
######################################################################
    RRDTool::OO 0.35
######################################################################

NAME
    RRDTool::OO - Object-oriented interface to RRDTool

SYNOPSIS
        use RRDTool::OO;

            # Constructor     
        my $rrd = RRDTool::OO->new(
                     file => "myrrdfile.rrd" );

            # Create a round-robin database
        $rrd->create(
             step        => 1,  # one-second intervals
             data_source => { name      => "mydatasource",
                              type      => "GAUGE" },
             archive     => { rows      => 5 });

            # Update RRD with sample values, use current time.
        for(1..5) {
            $rrd->update($_);
            sleep(1);
        }

            # Start fetching values from one day back, 
            # but skip undefined ones first
        $rrd->fetch_start();
        $rrd->fetch_skip_undef();

            # Fetch stored values
        while(my($time, $value) = $rrd->fetch_next()) {
             print "$time: ", 
                   defined $value ? $value : "[undef]", "\n";
        }

            # Draw a graph in a PNG image
        $rrd->graph(
          image          => "mygraph.png",
          vertical_label => 'My Salary',
          start          => time() - 10,
          draw           => {
              type   => "area",
              color  => '0000FF',
              legend => "Salary over Time",
          }
        );

            # Same using rrdtool's graphv
        $rrd->graphv(
          image          => "mygraph.png",
          [...]
        };

DESCRIPTION
    "RRDTool::OO" is an object-oriented interface to Tobi Oetiker's round
    robin database tool *rrdtool*. It uses *rrdtool*'s "RRDs" module to get
    access to *rrdtool*'s shared library.

    "RRDTool::OO" tries to marry *rrdtool*'s database engine with the
    dwimminess and whipuptitude Perl programmers take for granted. Using
    "RRDTool::OO" abstracts away implementation details of the RRD engine,
    uses easy to memorize named parameters and sets meaningful defaults for
    parameters not needed in simple cases. For the experienced user,
    however, it provides full access to *rrdtool*'s API (if you find a
    feature that's not implemented, let me know).

  FUNCTIONS
    *my $rrd = RRDTool::OO->new( file => $file )*
        The constructor hooks up with an existing RRD database file $file,
        but doesn't create a new one if none exists. That's what the
        "create()" methode is for. Returns a "RRDTool::OO" object, which can
        be used to get access to the following methods.

    *$rrd->create( ... )*
        Creates a new round robin database (RRD). A RRD consists of one or
        more data sources and one or more archives:

            $rrd->create(
                 step        => 60,
                 data_source => { name      => "mydatasource",
                                  type      => "GAUGE" },
                 archive     => { rows      => 5 });

        This defines a RRD database with a step rate of 60 seconds in
        between primary data points. Additionally, the RRD start time can be
        specified by specifying a "start" parameter.

        It also sets up one data source named "my_data_source" of type
        "GAUGE", telling *rrdtool* to use values of data samples as-is,
        without additional trickery.

        And it creates a single archive with a 1:1 mapping between primary
        data points and archive points, with a capacity to hold five data
        points.

        The RRD's "step" parameter is optional, and will be set to 300
        seconds by *rrdtool* by default.

        In addition to the mandatory settings for "name" and "type",
        "data_source" parameter takes the following optional parameters:
        "min" (minimum input, defaults to "U"), "max" (maximum input,
        defaults to "U"), "heartbeat" (defaults to twice the RRD's step
        rate).

        Archives expect at least one parameter, "rows" indicating the number
        of data points the archive is configured to hold. If nothing else is
        set, *rrdtool* will store primary data points 1:1 in the archive.

        If you want to combine several primary data points into one archive
        point, specify values for "cpoints" (the number of points to
        combine) and "cfunc" (the consolidation function) explicitly:

            $rrd->create(
                 step        => 60,
                 data_source => { name      => "mydatasource",
                                  type      => "GAUGE" },
                 archive     => { rows      => 5,
                                  cpoints   => 10,
                                  cfunc     => 'AVERAGE',
                                });

        This will collect 10 data points to form one archive point, using
        the calculated average, as indicated by the parameter "cfunc"
        (Consolidation Function, CF). Other options for "cfunc" are "MIN",
        "MAX", and "LAST".

        If you're defining multiple data sources or multiple archives, just
        provide them in this manner:

               # Define the RRD
            my $rc = $rrd->create(
                step        => 60,
                data_source => { name      => 'load1',
                                 type      => 'GAUGE',
                               },
                data_source => { name      => 'load2',
                                 type      => 'GAUGE',
                               },
                archive     => { rows      => 5,
                                 cpoints   => 10,
                                 cfunc     => 'AVERAGE',
                                },
                archive     => { rows      => 5,
                                 cpoints   => 10,
                                 cfunc     => 'MAX',
                                },
            );

    *$rrd->update( ... ) *
        Update the round robin database with a new data sample, consisting
        of a value and an optional time stamp. If called with a single
        parameter, like in

            $rrd->update($value);

        then the current timestamp and the defined $value will be used. If
        "update" is called with a named parameter list like in

            $rrd->update(time => $time, value => $value);

        then the given timestamp $time is used along with the given value
        $value.

        When updating multiple data sources, use the "values" parameter
        (instead of "value") and pass an arrayref:

            $rrd->update(time => $time, values => [$val1, $val2, ...]);

        This way, *rrdtool* expects you to pass in the data values in
        exactly the same order as the data sources were defined in the
        "create" method. If that's not the case, then the "values" parameter
        also accepts a hashref, mapping data source names to values:

            $rrd->update(time => $time, 
                         values => { $dsname1 => $val1, 
                                     $dsname2 => $val2, ...});

        "RRDTool::OO" will transform this automagically into "RRDTool's"
        *template* syntax.

    *$rrd->updatev( ... )*
        This is identical to "update", but uses rrdtool's updatev function
        internally. The only difference is when using the "print_results"
        method described below, which then contains additional information.

    *$rrd->fetch_start( ... )*
        Initializes the iterator to fetch data from the RRD. This works
        nicely without any parameters if your archives are using a single
        consolidation function (e.g. "MAX"). If there's several archives in
        the RRD using different consolidation functions, you have to specify
        which one you want:

            $rrd->fetch_start(cfunc => "MAX");

        Other options for "cfunc" are "MIN", "AVERAGE", and "LAST".

        "fetch_start" features a number of optional parameters: "start",
        "end" and "resolution".

        If the "start" time parameter is omitted, the fetch starts 24 hours
        before the end of the archive. Also, an "end" time can be specified:

            $rrd->fetch_start(start => time()-10*60,
                              end   => time());

        The third optional parameter, "resolution" defaults to the highest
        resolution available and can be set to a value in seconds,
        specifying the time interval between the data samples extracted from
        the RRD. See the "rrdtool fetch" manual page for details.

        Development note: The current implementation fetches *all* values
        from the RRA in one swoop and caches them in memory. This might
        change in the future, to cache only the last timestamp and keep
        fetching from the RRD with every "fetch_next()" call.

    *$rrd->fetch_skip_undef()*
        *rrdtool* doesn't remember the time the first data sample went into
        the archive. So if you run a *rrdtool fetch* with a start time of 24
        hours ago and you've only submitted a couple of samples to the
        archive, you'll see many "undef" values.

        Starting from the current iterator position (or at the specified
        "start" time immediately after a "fetch_start()"),
        "fetch_skip_undef()" will skip all "undef" values in the RRA and
        positions the iterator right before the first defined value. If all
        values in the RRA are undefined, the a following
        "$rrd->fetch_next()" will return "undef".

    *($time, $value, ...) = $rrd->fetch_next()*
        Gets the next row from the RRD iterator, initialized by a previous
        call to "$rrd->fetch_start()". Returns the time of the archive point
        along with all values as a list.

        Note that there might be more than one value coming back from
        "fetch_next" if the RRA defines more than one datasource):

            I<($time, @values_of_all_ds) = $rrd-E<gt>fetch_next()>

        It is not possible to fetch only a specific datasource, as rrdtool
        doesn't provide this.

    *($time, $value, ...) = $rrd->fetch_next()*
    *$rrd->graph( ... )*
        If there's only one data source in the RRD, drawing a nice graph in
        an image file on disk is as easy as

            $rrd->graph(
              image          => $image_file_name,
              vertical_label => 'My Salary',
              draw           => { thickness => 2,
                                  color     => 'FF0000',
                                  legend    => 'Salary over Time',
                                },
            );

        This will assume a start time of 24 hours before now and an end time
        of now. Specify "start" and "end" explicitly to be clear:

            $rrd->graph(
              image          => $image_file_name,
              vertical_label => 'My Salary',
              start          => time() - 24*3600,
              end            => time(),
              draw           => { thickness => 2,
                                  color     => 'FF0000',
                                  legend    => 'Salary over Time',
                                },
            );

        As always, "RRDTool::OO" will pick reasonable defaults for
        parameters not specified. The values for data source and
        consolidation function default to the first values it finds in the
        RRD. If there are multiple datasources in the RRD or multiple
        archives with different values for "cfunc", just specify explicitly
        which one to draw:

            $rrd->graph(
              image          => $image_file_name,
              vertical_label => 'My Salary',
              draw           => {
                thickness => 2,
                color     => 'FF0000',
                dsname    => "load",
                cfunc     => 'MAX'},
            );

        If "draw" doesn't define a "type", it defaults to "line". If you
        don't want to define a type (because the graph shouldn't be drawn),
        use "type => "hidden"". Other values are "area" for solid colored
        areas. The "stack" type (for graphical values stacked on top of each
        other) has been deprecated sind rrdtool-1.2, but RRDTool::OO still
        supports it by transforming it into an 'area' type with a 'stack'
        option.

        And you can certainly have more than one graph in the picture:

            $rrd->graph(
              image          => $image_file_name,
              vertical_label => 'My Salary',
              draw           => {
                type      => 'area',
                color     => 'FF0000', # red area
                dsname    => "load",
                cfunc     => 'MAX'},
              draw        => {
                type      => 'area',
                stack     => 1,
                color     => '00FF00', # a green area stacked on top of the red one 
                dsname    => "load",
                cfunc     => 'AVERAGE'},
            );

        Graphs may assemble data from different RRD files. Just specify
        which file you want to draw the data from, using "draw":

            $rrd->graph(
              image          => $image_file_name,
              vertical_label => 'Network Traffic',
              draw           => {
                file      => "file1.rrd",
                legend    => "First Source",
              },
              draw        => {
                file      => "file2.rrd",
                type      => 'area',
                stack     => 1,
                color     => '00FF00', # a green area stacked on top of the red one 
                dsname    => "load",
                legend    => "Second Source",
                cfunc     => 'AVERAGE'
              },
            );

        If a "file" parameter is specified per "draw", the defaults for
        "dsname" and "cfunc" are fetched from this file, not from the file
        that's attached to the "RRDTool::OO" object $rrd used.

        Graphs may also consist of algebraic calculations of previously
        defined graphs. In this case, graphs derived from real data sources
        need to be named, so that subsequent "cdef" definitions can refer to
        them and calculate new graphs, based on the previously defined
        graph:

            $rrd->graph(
              image          => $image_file_name,
              vertical_label => 'Network Traffic',
              draw           => {
                type      => 'line',
                color     => 'FF0000', # red line
                dsname    => 'load',
                name      => 'firstgraph',
                legend    => 'Unmodified Load',
              },
              draw        => {
                type      => 'line',
                color     => '00FF00', # green line
                cdef      => "firstgraph,2,*",
                legend    => 'Load Doubled Up',
              },
            );

        Note that the second "draw" doesn't refer to a datasource "dsname"
        (nor does it fall back to the default data source), but defines a
        "cdef", performing calculations on a previously defined draw named
        "firstgraph". The calculation is specified using RRDTool's reverse
        polish notation, where instructions are separated by commas
        ("firstgraph,2,*" simply multiplies "firstgraph"'s values by 2).

        On a global level, in addition to the "vertical_label" parameter
        shown in the examples above, "graph" offers a plethora of
        parameters:

        "vertical_label", "title", "start", "end", "x_grid", "y_grid",
        "alt_y_grid", "no_minor", "alt_y_mrtg", "alt_autoscale",
        "alt_autoscale_max", "base", "units_exponent", "units_length",
        "width", "height", "interlaced", "imginfo", "imgformat", "overlay",
        "unit", "lazy", "rigid", "lower_limit", "upper_limit",
        "logarithmic", "color", "no_legend", "only_graph",
        "force_rules_legend", "title", "step".

        Some options (e.g. "alt_y_grid") don't expect values, they need to
        be specified like

            alt_y_grid => undef

        in order to be passed properly to RRDTool.

        The "color" option expects a reference to a hash with various
        settings for the different graph areas: "back" (background),
        "canvas", "shadea" (left/top border), "shadeb" (right/bottom
        border), "grid", "mgrid" major grid, "font", "frame" and "arrow":

            $rrd->graph(
              ...
              color          => { back   => '#0e0e0e',
                                  arrow  => '#ff0000',
                                  canvas => '#eebbbb',
                                },
              ...
            );

        Fonts for various graph elements may be specified in "font" blocks,
        which must either name a TrueType font file or a PDF/Postscript font
        name. You may optionally specify a size and element name (defaults
        to DEFAULT, which to RRD means "use this font for everything).
        Example:

            font  => {
                name => "/usr/openwin/lib/X11/fonts/TrueType/GillSans.ttf",
                size => 16,
                element => "title"
            }

        Please check the RRDTool documentation for a detailed description on
        what each option is used for:

            http://people.ee.ethz.ch/~oetiker/webtools/rrdtool/manual/rrdgraph.html

        Sometimes it's useful to print max, min or average values of a given
        graph at the bottom of the chart or to STDOUT. That's what "gprint"
        and "print" options are for. They are printing variables which are
        defined as "vdef"s somewhere else:

            $rrd->graph(
              image          => $image_file_name,
                  # Real graph
              draw           => {
                name      => "first_draw",
                dsname    => "load",
                cfunc     => 'MAX'
              },

                # vdef for calculating average of real graph
              draw           => {
                type      => "hidden",
                name      => "average_of_first_draw",
                vdef      => "first_draw,AVERAGE"
              },

              gprint         => {
                draw      => 'average_of_first_draw',
                format    => 'Average=%lf',
              },
            );

        The "vdef" performs a calculation, specified in RPN notation, on a
        real graph, which it refers to. It uses a hidden graph for this.

        The "gprint" option then refers to the "vdef" virtual graph and
        prints "Average=x.xx" at the bottom of the graph, showing what the
        average value of graph "first_draw" is.

        To write comments to the graph (like gprints, but with no associated
        RRD data source) use "comment", like this:

            $rrd->graph(
              image          => $image_file_name,
              draw           => {
                name      => "first_draw",
                dsname    => "load",
                cfunc     => 'MAX'},
              comment        => "Remember, 83% of all statistics are made up",
            );

        Multiple comment lines can be specified in a single comment
        specification like this:

             comment => [ "All the king's horses and all the king's men\\n",
                          "couldn't put Humpty together again.\\n",
                        ],

        Vertical rules (lines) may be placed into the graph by using a
        "vrule" block like so:

               vrule => { time => time()-3600, }

        These can be useful for indicating when the most recent day on the
        graph started, for example.

        vrules can have a color specification (they default to black) and
        also an optional legend string specified:

              vrule => { time => $first_thing_today,
                         color => "#0000ff",
                         legend => "When we crossed midnight"
                       },

        hrules can have a color specification (they default to black) and
        also an optional legend string specified:

              hrule => { value => $numeric_value,
                         color => "#0000ff",
                         legend => "a static line at your value"
                       },

        Horizontal rules can be added by using a "line" block like in

            line => { 
                value   => "fixed num value or draw name",
                color   => "#0000ff",
                legend  => "a blue horizontal line",
                width   => 120,
                stack   => 1,
            }

        If instead of a horizontal line, a rectangular area is supposed to
        be added to the graph, use an "area" block:

            area => { 
                value   => "fixed num value or draw name",
                color   => "#0000ff",
                legend  => "a blue horizontal line",
                stack   => 1,
            }

        The "graph" method can also generate tickmarks (vertical lines) for
        every defined value, using the "tick" option:

            tick => {
                draw    => "drawname",
                color   => "#0000ff",
                legend  => "a blue horizontal line",
                stack   => 1,
            }

        The graph may be shifted relative to the time axis:

            shift => {
                draw    => "drawname",
                offset  => $offset,
            }

    *$rrd->graphv( ... )*
        This is identical to "graph", but uses rrdtool's graphv function
        internally. The only difference is when using the "print_results"
        method described below, which then contains additional information.
        Be aware that rrdtool 1.3 is required for "graphv" to work.

    *$rrd->dump()*
        *Available as of rrdtool 1.0.49*.

        Dumps the RRD in XML format to STDOUT. If you want to dump it into a
        file instead, do this:

            my $pid;

            unless ($pid = open DUMP, "-|") {
              die "Can't fork: $!" unless defined $pid;
              $rrd->dump();
              exit 0;
            }

            waitpid($pid, 0);

            open OUT, ">out";
            print OUT $_ for <DUMP>;
            close OUT;

    *my $hashref = $rrd->xport(...)*
        Feed a perl structure with RRA data (Cf. rrdxport man page).

            my $results = $rrd->xport(
                start => $start_time,
                end => $end_time ,
                step => $step,
                def => [{
                    vname => "load1_vname",
                    file => "foo",
                    dsname => "load1",
                    cfunc => "MAX",
                },
                {
                    vname => "load2_vname",
                    file => "foo",
                    dsname => "load2",
                    cfunc => "MIN",
                }],

                cdef => [{
                    vname => "load2_vname_multiply",
                    rpn => "load2_vname,2,*",
                }],

                xport => [{
                    vname => "load1_vname",
                    legend => "it_s_gonna_be_legend_",
                },
                {
                    vname => "load2_vname",
                    legend => "wait_for_it",
                },
                {
                    vname => "load2_vname_multiply",
                    legend => "___dary",
                }],
            );

            my $data = $results->{data};
            my $metadata = $results->{meta};

            print "### METADATA ###\n";
            print "StartTime: $metadata->{start}\n";
            print "EndTime: $metadata->{end}\n";
            print "Step: $metadata->{step}\n";
            print "Number of data columns: $metadata->{columns}\n";
            print "Number of data rows: $metadata->{rows}\n";
            print "Legend: ", join(", ", @{$metadata->{legend}}), "\n";

            print "\n### DATA ###\n";
            foreach my $entry (@$data) {
                my $entry_timestamp = shift(@$entry);
                print "[$entry_timestamp] ", join(" ", @$entry), "\n";
            }

    *my $hashref = $rrd->info()*
        Grabs the RRD's meta data and returns it as a hashref, holding a map
        of parameter names and their values.

    *my $time = $rrd->last()*
        Return the RRD's last update time.

    *$rrd->restore(xml => "file.xml")*
        *Available as of rrdtool 1.0.49*.

        Restore a RRD from a "dump". The "xml" parameter specifies the name
        of the XML file containing the dump. If the optional flag
        "range_check" is set to a true value, "restore" will make sure the
        values in the RRAs do not exceed the limits defined for the
        different datasources:

            $rrd->restore(xml => "file.xml", range_check => 1);

    *$rrd->tune( ... )*
        Alter a RRD's data source configuration values:

                # Set the heartbeat of the RRD's only datasource to 100
            $rrd->tune(heartbeat => 100);

                # Set the minimum of DS 'load' to 1
            $rrd->tune(dsname => 'load', minimum => 1);

                # Set the maximum of DS 'load' to 10
            $rrd->tune(dsname => 'load', maximum => 10);

                # Set the type of DS 'load' to AVERAGE
            $rrd->tune(dsname => 'load', type => 'AVERAGE');

                # Set the name of DS 'load' to 'load2'
            $rrd->tune(dsname => 'load', name => 'load2');

    *$rrd->error_message()*
        Return the message of the last error that occurred while interacting
        with "RRDTool::OO".

  Aberrant behavior detection
    RRDTool supports aberrant behavior detection (ABD), which takes a data
    source, stuffs its values into a special RRA, smoothes the data stream,
    tries to predict future values and triggers an alert if actual values
    are way off the predicted values.

    Using a fairly elaborate algorithm not only allows it to find out if a
    data source produces a value that exceeds a certain fixed threshold. The
    algorithm constantly adapts its parameters to the input data and acts
    dynamically on slowly changing values.

    The "alpha" parameter specifies the baseline and lies between 0 and 1.
    Values close to 1 specify that most recent values have the most weight
    on the prediction, whereas values close to 0 indicate that past values
    carry higher weight.

    On top of that, ABD can deal with data input that displays continuously
    rising values (slope). The "beta" parameters, again between 0 and 1,
    specifies whether past values or more recent values carry the most
    weight.

    And, furthermore, it deals with seasonal cycles, so it won't freak out
    if there's a daily peak at noon. The "gamma" parameter indicates this,
    if you don't specify it, it defaults to the value of "alpha".

    In the easiest case, an RRA with aberrant behavior detection can be
    created like

            # Create a round-robin database
        $rrd->create(
             step        => 1,  # one-second intervals
             data_source => { name      => "mydatasource",
                              type      => "GAUGE" },
             hwpredict   => { rows => 3600,
                            },
        );

    where "alpha" and "beta" default to 0.5, and the "seasonal_period"
    defaults to 1/5 of the rows number.

    "rows" is the number of primary data points that are stored in the RRA

Changes
0.35  (07/20/2014)
    (ms) Fix for [rt.cpan.org #97322], now rounding results to cope with
         floating point inaccuracies in rrdtool. Reported by Andreas König.

0.34  (06/26/2014)
    (ms) Added disable_rrdtool_tag option as suggested by Sun Guonian.

0.33  (07/08/2013)
    (ms) Florian Eckert reported that graph() modified 2nd level
         entries of the options array passed to it. Used Storable::clone
         in OO.pm to make a deep copy first.

0.32  (03/06/2012)
    (ms) [rt.cpan.org #63351] Applied modified patch by Jonas Wagner to
         fix stacked graphs with no legend.
    (ms) Implements RRDs::xport and returns a Perl structure instead of 
         basic xport array (Remi Ferrand, Fabien Wernli).

0.31  (05/26/2010)
    (ms) [rt.cpan.org #54870] Fixed typos/pod errors reported by
         Salvatore Bonaccorso.
    (ms) Fixed overwriting input parameters in create(), bug reported
         by Erik Wasser.

0.30  (02/21/2010)
    (ms) [RT 53961] Worked around rrdtool-1.3.5 inaccuracy problem by
         skipping certain tests for rrdtool <= 1.4.

0.29  (02/20/2010)
    (ms) Added github repository link to Makefile.PL
    (ms) Fixed documentation for fetch_next().
    (ms) [RT 54544] Tom Regner added support for updatev().

0.28 (11/05/2009)
    (ms) Applied patch by Lyle Brooks, adding optional step, start, and
         end parameters to "draw".

0.27 (10/11/2009)
    (ms) Slaven Rezic reported huge memory consumption by the test suite,
         turned out to be fetch_start() call spanning several years. 
         Fixed by defining the end time as well as the start time.

0.26 (09/27/2009)
    (ms) [RT 32601] Using prompt() instead of manually asking for input 
         in Makefile.PL (by Fabien Wernli)
    (ms) [RT 32046] Calling setlocale() to fix error message check in 
         foreign locales (by Fabien Wernli)
    (ms) Implemented graphv method. This works just like graph() but uses 
         rrdtool's graphv internally, giving access to additional 
         information using print_results() method (by Fabien Wernli)
    (ms) Improved RRDs version testing (by Fabien Wernli).

0.25 (04/25/2009)
    (ms) Applied patch by Christian Dengler, adding the 'hrule' graph
         parameter and increasing the sleep time in 002Pod.t to two
         secs to prevent system clock inacurracies from causing
         test failures.

0.24 (04/02/2009)
    (ms) Added support for capturing PRINT output as suggested by
         Peter Mistich. print_output() now returns a ref to an array
         with the output of the last graph() command's PRINT output.

0.23 (02/05/2009)
    (ms) update() now accepts DateTime objects as well (suggested by 
         Jay Buffington)
    (ms) [rt.cpan.org #43011] Fixed rounding error on 64-bit systems
         reported by Todd E. Rinaldo

0.22 (05/20/2008)
    (ms) Added aberrant behavior detection (hwpredict) with documentation
         and test cases.

0.21 (01/23/2008)
    (ms) Provided backwards compatibility to rrdtool-1.0 for 
         draw/type=stack, which now translates to an area with the 
         :STACK option.

0.20 (01/22/2008)
    (ms) Bas van der Veen reported that the rrdtool download link 
         changed to http://oss.oetiker.ch/rrdtool/pub/rrdtool.tar.gz,
         adapted Makefile.PL.
    (ms) No longer expecting 'Permission denied' to support localized
         OSes.

0.19 (05/12/2007)
    (ms) Added 'strict' mode (defaults to true).
    (ms) Added latest graph parameters from the latest rrdtool release.

0.18 (05/12/2007)
    (ms) Added 'dry mode' patch by Jacquelin Charbonnel.

0.17 (09/10/2006)
    (ms) Changed build command in Makefile.PL to adapt to new
         rrdtool-1.2 directory hierarchy, as suggested by Zac Israel.

0.16 (08/23/2006)
    (ms) Added 'base' option to graph method, as suggested by 
         Anton Shevchenko.

0.15 (07/23/2006)
    (ms) Fixed bug with multiple comment lines, of which only the first
         one was printed in the graph. It was introduced in 0.14. 
         Thanks to Steve van der Burg for reporting and a patch.

0.14 (04/28/2006)
    (ms) Order of prints, legends, gprints etc. now preserved, see
         http://lists.ee.ethz.ch/rrd-users/msg11161.html

0.13 (08/07/2005)
    (ms) Added line, area, tick, and shift to graph method.
    (ms) Added vdef to graph.
    (ms) More checks on rrdtool installation, now requiring 1.2.x.

0.12 (07/07/2005)
    (ms) Steve van der Burg added vrule, comment, and font options
    (ms) Makefile.PL dependencies version number changed from undef to 0
    (ms) Added 'rigid' option to graph.

0.11 (01/05/2005)
    (ms) Added suggestion by Richard Lippmann <lippmann@sprachtheorie.de>
         to allow for invisible graphs, just used as base for later 
         'cdef' calculations. Since leaving 'type' out will have
         RRDTool::OO default to 'line', type => 'hidden' is used to
         indicate the graph shouldn't be drawn. 
    (ms) Added gprint/print options for graphs

0.10 (11/20/2004)
    (ms) Added legends to graph()
    (ms) Maxence Gerbedoen <maxxenceg@wanadoo.fr> suggested a patch to 
         add 'cdef' functionality. Added slightly differently.

0.09 (10/08/2004)
    (ms) Fixed bug [cpan #7897]. new() now accepts a raise_error setting.

0.08 (09/24/2004)
    (ms) Added documentation for multiple data sources and archives.
    (ms) Martin Kurahaupo <martin@kurahaupo.gen.nz> added a patch
         to prevent the test suite from failing when run as root
         because a test file can't be write-protected.
    (ms) Chris Fedde <cfedde@denver.pm.org> provided a patch to 
         structure ::info's output more perl-like. Got input from
         perlmonks (thanks nobull!) on how to efficiently transform
         RRDTools info output into Perl data structures.

0.07 (08/15/2004)
    (ms) doc and test fixes for new rrdtool release 1.0.49. dump()
         and restore() are now supported.

0.06 (06/21/2004)
    (ms) added 'start' param in create() docs

eg/graph.pl
#!/usr/bin/perl
############################################################
# Create a sample graph
# Mike Schilli <mschilli1@aol.com>, 2004
############################################################

use strict;
use warnings;

use RRDTool::OO;
use Log::Log4perl qw(:easy);
Log::Log4perl->easy_init($DEBUG);

my $DB  = "example.rrd";
my $IMG = "example.png";

my $rrd = RRDTool::OO->new(file => $DB);

    # Use a reproducable point in time
my $start_time     = 1080460200;
my $nof_iterations = 40;
my $end_time       = $start_time + $nof_iterations * 60;

   # Define the RRD
my $rc = $rrd->create(
    start       => $start_time - 10,
    step        => 60,
    data_source => { name      => 'load1',
                     type      => 'GAUGE',
                   },
    data_source => { name      => 'load2',
                     type      => 'GAUGE',
                   },
    archive     => { rows     => 50,
                   },
);

   # Pump in values
for(0..$nof_iterations) {
    my $time = $start_time + $_ * 60;
    my $value = 2 + $_ * 0.1;

    $rrd->update(
        time   => $time, 
        values => { 
            load1 => $value,
            load2 => $value+1,
        }
    );
}

   # Draw a graph of two different data sources,
   # stacked on top of each other
$rrd->graph(
    image          => $IMG,
    vertical_label => 'A Nice Area Graph',
    start          => $start_time,
    end            => $start_time + $nof_iterations * 60,
    width          => 700,
    height         => 300,
    color          => { back   => '#eeeeee',
                        arrow  => '#ff0000',
                        canvas => '#eebbbb',
                      },
        # First graph
    draw           => {
        name      => 'some_stupid_draw',
        type      => "area",
        color     => '0000ff',
        legend    => 'first legend',
        dsname    => 'load1',
    },
        # Second graph
    draw           => {
        type      => "stack",
        color     => '00ff00',
        dsname    => 'load2',
        legend    => 'second legend',
    },

#    gprint        => {
#        draw      => 'some_stupid_draw',
#        format    => 'avg=%lf',
#        #cfunc     => 'MIN',
#    },
);

print "$IMG ready.\n";
